lib/graph_gateway.go

```diff
@@ -240,606 +218,1135 @@ func registerGraphGatewayMetrics() *GraphGatewayMetrics {
 		Help:      "Tracks the size of range requests.",
 		Buckets:   prometheus.ExponentialBuckets(256*1024, 2, 10), // From 256KiB to 100MiB
 	})
-	prometheus.MustRegister(bytesRangeSizeMetric)
+	registerer.MustRegister(bytesRangeSizeMetric)
 
 	return &GraphGatewayMetrics{
 		contextAlreadyCancelledMetric,
 		carFetchAttemptMetric,
 		carBlocksFetchedMetric,
-		blockRecoveryAttemptMetric,
 		carParamsMetric,
 		bytesRangeStartMetric,
 		bytesRangeSizeMetric,
 	}
 }
 
-func (api *GraphGateway) getRootOfPath(path string) string {
-	pth, err := ipfspath.ParsePath(path)
-	if err != nil {
-		return path
-	}
-	if pth.IsJustAKey() {
-		return pth.Segments()[0]
-	} else {
-		return pth.Segments()[1]
+func paramsToString(params gateway.CarParams) string {
+	paramsBuilder := strings.Builder{}
+	paramsBuilder.WriteString("dag-scope=")
+	paramsBuilder.WriteString(string(params.Scope))
+	if params.Range != nil {
+		paramsBuilder.WriteString("&entity-bytes=")
+		paramsBuilder.WriteString(strconv.FormatInt(params.Range.From, 10))
+		paramsBuilder.WriteString(":")
+		if params.Range.To != nil {
+			paramsBuilder.WriteString(strconv.FormatInt(*params.Range.To, 10))
+		} else {
+			paramsBuilder.WriteString("*")
+		}
 	}
+	return paramsBuilder.String()
 }
 
-/*
-Implementation iteration plan:
+func (api *GraphGateway) fetchCAR(ctx context.Context, path gateway.ImmutablePath, params gateway.CarParams, cb DataCallback) error {
+	escapedPath := url.PathEscape(path.String())
+	escapedPath = strings.ReplaceAll(escapedPath, "%2F", "/")
+	paramsStr := paramsToString(params)
+	urlWithoutHost := fmt.Sprintf("%s?%s", escapedPath, paramsStr)
```

Comment:
Passing a `url` object rather than a string here seems like a good idea to me. We should be clear about the contract of what's allowed/not in the URL though.

For example, if it's valid/invalid to put `format=raw` in the params or headers or if the `Fetch` function implies CAR. While it's somewhat obviously true now it doesn't have to be that way so the docs should clarify either way.

Reply:
@aschmahmann  I've pushed small refactor in https://github.com/ipfs/bifrost-gateway/pull/160/commits/982535a97b046017cc320ddb08a9fcb44c2a1b94 that does not change the `fetcher` API (we still pass URL as string), but fixes the way we encode content paths.  I also added explicit `format=car` to make it easier to debug and test. 

I don't want to block this PR on the fetcher API cleanup  unless we really need to include it, and it does not seem to me like we need to solve problem at hand: it seems to work as expected already.

Reply:
@lidel yeah, I don't think we need to block on a refactor unless it's hindering functionality in some way. If this is enough for now and the refactor is cleanup + reducing headaches here then can definitely be a follow up.

Updating the top-level post with the areas currently under evaluation/investigation before merging.